### ISSUE TITLE
Fix `declaration-property-value-no-unknown` false positives for descriptors

### ISSUE THREAD

**Issue by FloEdelmann** [stylelint/stylelint#6663]
### What steps are needed to reproduce the bug?

```css
@font-face {
    font-family: 'AwesomeVariableFont';
    font-style: normal;
    font-weight: 100 700;
    src: url('./fonts/AwesomeVariableFont.woff2') format('woff2-variations');
}
```

For variable fonts, the `font-weight` descriptor accepts the minimum and maximum weights supported by the font:

* [MDN: `@font-face`](https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face#font-weight)
* [MDN: `@font-face`'s `font-weight` descriptor](https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/font-weight) (compare also with [MDN: `font-weight` property](https://developer.mozilla.org/en-US/docs/Web/CSS/font-weight))

### What Stylelint configuration is needed to reproduce the bug?

```yaml
rules:
  declaration-property-value-no-unknown: true
```

### How did you run Stylelint?

npx stylelint style.css

### Which version of Stylelint are you using?

15.1.0

### What did you expect to happen?

No problems to be reported.

### What actually happened?

The following problem was reported:

```
style.css
 4:22  ✖  Unexpected unknown value "700" for property "font-weight"  declaration-property-value-no-unknown
```

### Does the bug relate to non-standard syntax?

No, only native CSS syntax is involved.

**Comment by ybiquitous** [comment]
@FloEdelmann Thanks for the report with the detailed reproduction. We can reproduce the problem on the [demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIACAzCA7ALgWhQQzBgAJgAddIotLbAdxgEsBzAC03iIEYAGbogdl4BucgF9yIADQgAzgE8seAB4JwMmVPAYUzVWQpFSIAE4BXADYwZRjvsqUjsMObzG8mBhmwAHYxG8wxphy2ABueOamMNjoENim6ADWsbToNkSYZjDklOLooiCiQA).

Perhaps, CSSTree (the backend of `declaration-property-value-no-unknown`) may have the same problem. See the [CSSTree demo](https://csstree.github.io/docs/validator.html):

<details>
<summary>Screenshot</summary>

<img width="875" alt="Screenshot 2023-02-15 at 23 03 56" src="https://user-images.githubusercontent.com/473530/219048927-114b287d-134a-4460-8f0d-fb39a126cfb4.png">

</details>

@jeddy3 Can we leave the resolution to CSSTree, or solve it ourselves?

**Comment by jeddy3** [comment]
Let's resolve it upstream (either in [CSSTree](https://github.com/csstree/csstree/blob/master/data/patch.json) or [mdn-data](https://github.com/mdn/data/), whichever is appropriate).

As a general rule of thumb for this rule, let's point the reporter upstream especially when the issue can be worked around by extending the syntax. We can keep the issue open here as a pointer for other users, until the upstream fix happens.

@FloEdelmann I've labelled the issue as ready to implement. Please consider contributing upstream if you have time.

You can also temporarily work around the issue by [extending the syntax](https://stylelint.io/user-guide/rules/declaration-property-value-no-unknown/#propertiessyntax--property-syntax-) to allow two value: 

```json
{
  "rules": {
    "declaration-property-value-no-unknown": [true, { "propertiesSyntax": {
      "font-weight": "| <font-weight-absolute>{2}"
    }}]
  }
}
```

[Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIACAzCA7ALgWhQQzBgAJgAddIotLbAdxgEsBzAC03iIEYAGbogdl4BucgF9yIADQgAzgE8seAB4JwMmVPAYUzVWQpFSIAE4BXADYwZRjvsqUjsMObzG8mBhmwAHYxG8wxphy2ABueOamMNjoENim6ADWsbToNkQA2phmMJIkhiC+-oEeVgDKCpjK6Xb2DiDUOPTMbOlGAD5EADyNdIysOHgARjIQkZgwAHzAAEziIOR1oqIAuotE4uiiIKJAA).

[Syntax reference](https://csstree.github.io/docs/syntax/#Property:font-weight).

**Comment by ybiquitous** [comment]
@jeddy3 I understood well, thank you! 😄

**Comment by FloEdelmann** [comment]
The `font-weight` descriptor looks correct in https://github.com/mdn/data/: https://github.com/mdn/data/blob/678604eeb7cad0c1e5d637fa6918922be1b24be3/css/at-rules.json#L206-L214
I created an issue in [CSSTree](https://github.com/csstree/csstree/blob/master/data/patch.json?rgh-link-date=2023-02-15T16%3A58%3A46Z):

* https://github.com/csstree/csstree/issues/247

**Comment by FloEdelmann** [comment]
Maybe it is an issue in Stylelint after all? I can see that CSSTree has a separate `font-weight` descriptor for the `@font-face` at-rule, which accepts the format: https://csstree.github.io/docs/syntax/#AtruleDescriptor:font-weight

**Comment by jeddy3** [comment]
@FloEdelmann Thank you for digging into this some more!

It seems likely we'll need to make changes to rule as it doesn't account for `font-weight` having a different syntax depending on the context. 

As a descriptor ([a declaration within a `@font-face`](https://w3c.github.io/csswg-drafts/css-fonts/#font-face-rule)) it can accept two values:

```css
@font-face { 
  font-weight: 100 700;
}
```

As a declaration within a rule set it can't:

```css
a {
  font-weight: 700;
}
```

As such, extending the syntax in https://github.com/stylelint/stylelint/issues/6663#issuecomment-1431685951 will introduce false negatives for rule sets.

We'll need to delve deeper. Perhaps CSSTree has a way to specify the context when parsing declarations? So that the descriptor syntax is used for declarations within `@font-face` and the property syntax otherwise.

Incidentally, do we know if any other properties behave this way or is it only `font-weight`?

**Comment by FloEdelmann** [comment]
It seems there are quite a lot actually:

* `speak-as`: [`@counter-style` descriptor](https://csstree.github.io/docs/syntax/#AtruleDescriptor:speak-as) vs. [property](https://csstree.github.io/docs/syntax/#Property:speak-as)
* `font-family`: [`@font-face` descriptor](https://csstree.github.io/docs/syntax/#AtruleDescriptor:font-family) vs. [property](https://csstree.github.io/docs/syntax/#Property:font-family)
* `font-feature-settings`: [`@font-face` descriptor](https://csstree.github.io/docs/syntax/#AtruleDescriptor:font-feature-settings) vs. [property](https://csstree.github.io/docs/syntax/#Property:font-feature-settings), although they have the same syntax
* `font-variation-settings`: [`@font-face` descriptor](https://csstree.github.io/docs/syntax/#AtruleDescriptor:font-variation-settings) vs. [property](https://csstree.github.io/docs/syntax/#Property:font-variation-settings), although they have the same syntax
* `font-stretch`: [`@font-face` descriptor](https://csstree.github.io/docs/syntax/#AtruleDescriptor:font-stretch) vs. [property](https://csstree.github.io/docs/syntax/#Property:font-stretch), although they have the same syntax
* `font-style`: [`@font-face` descriptor](https://csstree.github.io/docs/syntax/#AtruleDescriptor:font-style) vs. [property](https://csstree.github.io/docs/syntax/#Property:font-style)
* `font-weight`: [`@font-face` descriptor](https://csstree.github.io/docs/syntax/#AtruleDescriptor:font-weight) vs. [property](https://csstree.github.io/docs/syntax/#Property:font-weight)
* `font-variant`: [`@font-face` descriptor](https://csstree.github.io/docs/syntax/#AtruleDescriptor:font-variant) vs. [property](https://csstree.github.io/docs/syntax/#Property:font-variant), although they have the same syntax
* `src`: [`@font-face` descriptor](https://csstree.github.io/docs/syntax/#AtruleDescriptor:src) vs. [property](https://csstree.github.io/docs/syntax/#Property:src), although they have the same syntax
* `unicode-range`: [`@font-face` descriptor](https://csstree.github.io/docs/syntax/#AtruleDescriptor:unicode-range) vs. [property](https://csstree.github.io/docs/syntax/#Property:unicode-range), although they have the same syntax
* `height`: [`@viewport` descriptor](https://csstree.github.io/docs/syntax/#AtruleDescriptor:height) vs. [property](https://csstree.github.io/docs/syntax/#Property:height)
* `max-height`: [`@viewport` descriptor](https://csstree.github.io/docs/syntax/#AtruleDescriptor:max-height) vs. [property](https://csstree.github.io/docs/syntax/#Property:max-height)
* `max-width`: [`@viewport` descriptor](https://csstree.github.io/docs/syntax/#AtruleDescriptor:max-width) vs. [property](https://csstree.github.io/docs/syntax/#Property:max-width)
* `min-height`: [`@viewport` descriptor](https://csstree.github.io/docs/syntax/#AtruleDescriptor:min-height) vs. [property](https://csstree.github.io/docs/syntax/#Property:min-height)
* `min-width`: [`@viewport` descriptor](https://csstree.github.io/docs/syntax/#AtruleDescriptor:min-width) vs. [property](https://csstree.github.io/docs/syntax/#Property:min-width)
* `width`: [`@viewport` descriptor](https://csstree.github.io/docs/syntax/#AtruleDescriptor:width) vs. [property](https://csstree.github.io/docs/syntax/#Property:width)
* `zoom`: [`@viewport` descriptor](https://csstree.github.io/docs/syntax/#AtruleDescriptor:zoom) vs. [property](https://csstree.github.io/docs/syntax/#Property:zoom)

So I think a proper fix would need to take the surrounding at-rule into consideration.

**Comment by jeddy3** [comment]
@FloEdelmann Thank you for sharing these.

It looks like we will need a general solution that works for `@counter-style`, `@font-face` and `@viewport`.

If anyone has time, I believe the next step is to investigate whether CSSTree lets us specify the context when parsing declarations so that the correct syntax is used.

**Comment by FloEdelmann** [comment]
I've investigated and created a PR with the fix:

* https://github.com/stylelint/stylelint/pull/6669
